src/libAtomVM/nifs.c

```diff
@@ -3651,6 +3658,54 @@ static term nif_erlang_make_fun_3(Context *ctx, int argc, term argv[])
     return term_make_function_reference(module_term, function_term, arity_term, &ctx->heap);
 }
 
+static term nif_erlang_fun_info_2(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc);
+    term fun = argv[0];
+    term key = argv[1];
+
+    VALIDATE_VALUE(fun, term_is_fun);
+    VALIDATE_VALUE(key, term_is_atom);
+
+    term value;
+    switch (key) {
+        case MODULE_ATOM: {
+            term module_name;
+            term_get_function_mfa(fun, &module_name, NULL, NULL, ctx->global);
+            value = module_name;
+            break;
+        }
+        case NAME_ATOM: {
+            term function_name;
+            term_get_function_mfa(fun, NULL, &function_name, NULL, ctx->global);
+            value = function_name;
+            break;
+        }
+
+        case ARITY_ATOM: {
+            term arity;
+            term_get_function_mfa(fun, NULL, NULL, &arity, ctx->global);
+            value = arity;
+            break;
+        }
+
+        case TYPE_ATOM:
+            value = term_is_external_fun(fun) ? EXTERNAL_ATOM : LOCAL_ATOM;
+            break;
+
+        default:
+            RAISE_ERROR(BADARG_ATOM);
```

Comment:
I also tried implementing `env` today. The idea was to extract `n_freeze` terms after `arity` term in function heap pointer but I mostly got `nil` and invalid values. Do you know why it may not work?

Reply:
no idea, I need to look at the code. Anyway I think it makes sense implementing it only if we have code using it.

Reply:
I definitely need to look at the code.